test/unit-test/sigv4_utest.c

```diff
@@ -214,3 +532,725 @@ void test_SigV4_AwsIotDateToIso8601_Formatting_Error()
+    TEST_ASSERT_EQUAL( SigV4InvalidParameter, returnStatus );
+}
+
+/* TODO - Verify the generated signatures. */
+void test_SigV4_GenerateHTTPAuthorization_Happy_Paths()
+{
+    SigV4Status_t returnStatus;
+
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+
+    /* Attempt to generate the signature with a secret longer than the digest length. This
+     * causes the inner-most HMAC key of the signing key to be hashed down. */
+    creds.pSecretAccessKey = SECRET_KEY_LONGER_THAN_HASH_BLOCK;
+    creds.secretAccessKeyLen = SECRET_KEY_LONGER_THAN_HASH_BLOCK_LEN;
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+
+    /* Test case when the first step of the Signing Key generation HMAC operation encounters a key that is
+     * less than hash block in length. */
+    char secretKeyGeneratingBlockLenHMACKey[ SIGV4_HASH_MAX_BLOCK_LENGTH - strlen( SIGV4_HMAC_SIGNING_KEY_PREFIX ) ];
+    memset( secretKeyGeneratingBlockLenHMACKey, ( int ) 'K', sizeof( secretKeyGeneratingBlockLenHMACKey ) );
+    creds.pSecretAccessKey = secretKeyGeneratingBlockLenHMACKey;
+    creds.secretAccessKeyLen = sizeof( secretKeyGeneratingBlockLenHMACKey );
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+
+    /* S3 is the only service in which the URI is only encoded once. */
+    params.serviceLen = S3_SERVICE_NAME_LEN;
+    params.pService = S3_SERVICE_NAME;
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+
+    /* Coverage for the case where the service name has the same length as "s3". */
+    params.serviceLen = S3_SERVICE_NAME_LEN;
+    params.pService = SERVICE;
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+
+    /* Coverage for the null-terminated path. */
+    params.pHttpParameters->pPath = NULL_TERMINATED_PATH;
+    params.pHttpParameters->pathLen = NULL_TERMINATED_PATH_LEN;
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+
+    /* Coverage for double-encoded equals in query string value. */
+    params.pHttpParameters->pQuery = QUERY_VALUE_HAS_EQUALS;
+    params.pHttpParameters->queryLen = STR_LIT_LEN( QUERY_VALUE_HAS_EQUALS );
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
```

Comment:
So we never assert the actual signature matches what we expect. That's quite useful to have for a test, but I guess we've run out of time at this point.

Reply:
Or it would be good to have at least one test case that asserts the signature. Otherwise, this is really just a test of return codes for the sake of coverage...